README.md
# Blåhaj

Adds the popular IKEA plush to Caves of Qud 🦈

Blåhajar can be found in deep pools of convalessence.

Tested to work on game version 206.34.

[Source Code (GitHub)](https://github.com/librarianmage/Blahajar) \| [Workshop Page](https://steamcommunity.com/sharedfiles/filedetails/?id=2841450425) \| [My Caves of Qud Mods (Steam Workshop)](https://steamcommunity.com/profiles/76561198836298826/myworkshopfiles/?appid=333640)

## Acknowledgments

Thank you to ontoclasm for inspiration on the tile and Seagull for the description!

## Installation

Please refer to the Caves of Qud Wiki for [mod installation instructions](https://wiki.cavesofqud.com/wiki/Modding:Installing_a_mod).

## Development

In order for type-checking and auto-completion to work, place the game-generated `Mods.csproj` into the parent directory of this folder.

## License

I share the source of my mods with the intent that they can be useful for others. To this end, the code of this mod has been released with the Zero-Clause BSD (0BSD) license and the assets of this mod (namely, files in the `Assets/` and `Textures/` folders) have been released with the Creative Commons Zero v1.0 Universal (CC0-1.0) license.


CHANGELOG.md
# Changelog

## v1.3 — 2023-11-23
Tested to work on game version 206.34

* Refresh for the 7th Plague update

## v1.2 — 2023-08-14
Tested to work on game versions 204.100 / 205.54 (beta)

* Rebalance spawning frequency and weight
* Update names
* Refactor part code
* Add a Blåhaj to Neek's inventory

Meta:

* Add `CHANGELOG.md`
* Update icon
* Update mod description
* Update `LICENSE`

## v1.1 — 2022-08-03
Tested to work on game version 203.54

* Removes edibility from shonk stuffing (sorry)

## v1.0 — 2022-07-28
Tested to work on game version 203.54

* Initial release


blahaj.cs
using System;
using System.Linq;
using XRL.Rules;
using XRL.World.Capabilities;

namespace XRL.World.Parts
{
    /// <summary>
    ///   Has a chance of spawning a Blåhaj
    /// </summary>
    [Serializable]
    public class Books_BlahajSpawner : IPart
    {
        /// <summary>Spawning blueprint</summary>
        private const string Blueprint = "Books_Blahaj";

        /// <summary>Chances of spawning (percentage)</summary>
        public int Chance = 1;

        /// <summary>Chances of revealing when entering the ParentObject's cell (percentage)</summary>
        public int TrampleChance = 50;

        /// <summary>Difficulty of searching roll (INT)</summary>
        public int Search_Difficulty = 14;

        /// <summary>Whether to delete parent</summary>
        public bool DeleteParent = false;

        public override bool SameAs(IPart p) =>
            (p is Books_BlahajSpawner o)
            && o.Chance == Chance
            && o.Search_Difficulty == Search_Difficulty
            && o.DeleteParent == DeleteParent
            && base.SameAs(p);

        public override bool WantEvent(int ID, int cascade) =>
            base.WantEvent(ID, cascade)
            || ID == ObjectCreatedEvent.ID
            || ID == ObjectEnteredCellEvent.ID;

        public override bool HandleEvent(ObjectCreatedEvent E)
        {
            if (!Chance.in100())
            {
                ParentObject.RemovePart(this);

                if (DeleteParent)
                {
                    ParentObject.Obliterate(Silent: true);
                }
            }

            return base.HandleEvent(E);
        }

        public override bool HandleEvent(ObjectEnteredCellEvent E)
        {
            Cell currentCell = ParentObject.CurrentCell;

            if (
                E.Object.Blueprint != Blueprint
                && E.Object.IsCombatObject()
                && currentCell != null
                && TrampleChance.in100()
            )
            {
                Reveal(AdjacentCell: true);
            }

            return base.HandleEvent(E);
        }


        public override void Register(GameObject Object)
        {
            Object.RegisterPartEvent(this, "CustomRender");
            Object.RegisterPartEvent(this, "Searched");
            base.Register(Object);
        }



        public override bool FireEvent(Event E)
        {
            if (E.ID == "CustomRender")
            {
                if (
                    E.GetParameter("RenderEvent") is RenderEvent renderEvent
                    && (renderEvent.Lit == LightLevel.Radar
                        || renderEvent.Lit == LightLevel.LitRadar)
                )
                {
                    Reveal();
                }
            }
            else if (E.ID == "Searched")
            {
                GameObject Searcher = E.GetGameObjectParameter("Searcher");

                if (
                    Searcher.CurrentCell != ParentObject.CurrentCell
                    && Stat.MakeSave(Searcher, "Intelligence", Search_Difficulty)
                )
                {
                    Reveal(Searcher);
                }
            }
            return base.FireEvent(E);
        }

        public void Reveal(GameObject who = null, bool AdjacentCell = false)
        {
            who = who ?? The.Player;

            var cell = ParentObject.CurrentCell;

            if (AdjacentCell)
            {
                cell = ParentObject.CurrentCell
                    .YieldAdjacentCells(1)
                    .Where((c) => c.HasWadingDepthLiquid())
                    .GetRandomElement() ?? currentCell;
            }

            GameObject shonk = GameObject.create(Blueprint);
            shonk.MakeActive();
            cell.AddObject(shonk);

            XDidYToZ(
                Actor: who,
                Verb: "spot",
                Object: shonk,
                Extra: who.DescribeDirectionToward(shonk),
                EndMark: "!",
                ColorAsGoodFor: who,
                IndefiniteObject: true,
                UseVisibilityOf: shonk
            );

            if (
                Visible(shonk)
                && AutoAct.IsActive()
                && The.Player.IsRelevantHostile(shonk))
            {
                AutoAct.Interrupt(IndicateObject: shonk);
            }

            ParentObject.RemovePart(this);

            if (DeleteParent)
            {
                ParentObject.Obliterate(Silent: true);
            }
        }
    }

}


BlahajSpawner.cs
using System;
using System.Linq;
using XRL.Rules;

namespace XRL.World.Parts
{
    /// <summary>
    /// Gives the object a chance of spawning a hidden Blåhaj.
    /// </summary>
    /// <remarks>
    /// Object has a <see cref="SpawnChance"/>% chance of activating upon creation. When active, the spawning can be triggered under three circumstances:
    /// <list type="number">
    ///   <item>
    ///     <description>The cell is lit with <c>Omniscient</c>, <c>LitRadar</c>, or <c>Radar</c> light.</description>
    ///   </item>
    ///   <item>
    ///     <description>A creature enters the same cell, with <see cref="TrampleRevealChance"/>% chance.</description>
    ///   </item>
    ///   <item>
    ///     <description>The player makes an INT save of difficulty <see cref="SearchDifficulty"/> while searching.</description>
    ///   </item>
    /// </list>
    /// When the part fails to activate or successfully spawns, it will remove itself from its parent and optionally obliterate its parent if <see cref="ObliterateParent"/> is set.
    /// </remarks>
    /// <seealso cref="EelSpawn"/>
    /// <seealso cref="Hidden"/>
    /// <seealso cref="LightLevel"/>
    /// <seealso cref="Physics.Search"/>
    [Serializable]
    public class Books_BlahajSpawner : IPart
    {
        private const string BlahajBlueprint = "Books_Blahaj";

        /// <summary>
        /// The percentage chance that the part gets set up to spawn.
        /// </summary>
        public int SpawnChance = 1;

        /// <summary>
        /// The percentage chance that the object spawns when a creature enters the cell.
        /// </summary>
        public int TrampleRevealChance = 50;

        /// <summary>
        /// The difficulty of the INT roll used when searching.
        /// </summary>
        /// <seealso cref="Physics.Search"/>
        public int SearchDifficulty = 14;

        /// <summary>
        /// Whether to destroy the parent object after spawning.
        /// </summary>
        public bool ObliterateParent = false;

        public override void Initialize()
        {
            base.Initialize();

            ParentObject.pRender.CustomRender = true;
            ParentObject.ModIntProperty("CustomRenderSources", 1);
        }

        public override bool SameAs(IPart p) =>
            p is Books_BlahajSpawner o
            && o.SpawnChance == SpawnChance
            && o.TrampleRevealChance == TrampleRevealChance
            && o.SearchDifficulty == SearchDifficulty
            && o.ObliterateParent == ObliterateParent
            && base.SameAs(p);

        public override bool WantEvent(int ID, int cascade) =>
            base.WantEvent(ID, cascade)
            || ID == ObjectCreatedEvent.ID
            || ID == ObjectEnteredCellEvent.ID;

        public override bool HandleEvent(ObjectCreatedEvent E)
        {
            if (!SpawnChance.in100()) RemoveSelf();

            return base.HandleEvent(E);
        }

        public override bool HandleEvent(ObjectEnteredCellEvent E)
        {
            var currentCell = ParentObject.CurrentCell;

            if (E.Object.Blueprint != BlahajBlueprint
                && E.Object.IsCombatObject()
                && currentCell != null
                && TrampleRevealChance.in100())
            {
                Reveal(adjacentCell: true);
            }

            return base.HandleEvent(E);
        }


        public override void Register(GameObject Object)
        {
            Object.RegisterPartEvent(this, "CustomRender");
            Object.RegisterPartEvent(this, "Searched");

            base.Register(Object);
        }

        public override bool FireEvent(Event E)
        {
            if (E.ID == "CustomRender")
            {
                if (E.GetParameter("RenderEvent") is RenderEvent renderEvent
                    && (renderEvent.Lit == LightLevel.Omniscient
                        || renderEvent.Lit == LightLevel.Radar
                        || renderEvent.Lit == LightLevel.LitRadar))
                {
                    Reveal();
                }
            }
            else if (E.ID == "Searched")
            {
                var searcher = E.GetGameObjectParameter("Searcher");

                if (searcher.CurrentCell != ParentObject.CurrentCell
                    && Stat.MakeSave(searcher, "Intelligence", SearchDifficulty))
                {
                    Reveal(searcher);
                }
            }

            return base.FireEvent(E);
        }

        private void Reveal(GameObject finder = null, bool adjacentCell = false)
        {
            if (ParentObject.ModIntProperty("CustomRenderSources", -1, RemoveIfZero: true) == 0)
            {
                ParentObject.pRender.CustomRender = false;
            }

            finder ??= The.Player;

            var cell = currentCell;

            if (adjacentCell)
            {
                cell = cell
                    .YieldAdjacentCells(1)
                    .Where((c) => c.HasWadingDepthLiquid())
                    .GetRandomElement() ?? currentCell;
            }

            var shonk = GameObject.Create(BlahajBlueprint);
            shonk.MakeActive();
            cell.AddObject(shonk);

            XDidYToZ(
                Actor: finder,
                Verb: "spot",
                Object: shonk,
                Extra: finder.DescribeDirectionToward(shonk),
                EndMark: "!",
                ColorAsGoodFor: finder,
                IndefiniteObject: true,
                UseVisibilityOf: shonk
            );

            RemoveSelf();
        }

        private void RemoveSelf()
        {
            ParentObject.RemovePart(this);

            if (ObliterateParent) ParentObject.Obliterate(Silent: true);
        }
    }

}
